### PR TITLE
Upgrade toolchain to nightly-2022-09-13

### LINKED ISSUE #1615: Update rust toolchain version for Kani 0.11
We should update the rust toolchain version once every release. For 0.11, we will try to update to version nightly-2022-09-13.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Relevant changes to rustc:
  - https://github.com/rust-lang/rust/pull/101483: Change to intrinsics.
  - https://github.com/rust-lang/rust/pull/94075: Change to niche opt.
  - https://github.com/rust-lang/rust/pull/101101: Method rename.

### Resolved issues:

Fixes #1615

### Related RFC:

Optional #ISSUE-NUMBER.

### Call-outs:

I created a special case for the niche optimization to keep using `struct` instead of `union` for cases where the enumeration only has one non-ZST variant. Without this, the `kani` test suite takes more than 30min to finish. My guess is that it impacts a lot of `Option<>` and common enums out there. 

### Testing:

* How is this change tested? New tests

* Is this a refactor change? No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 1]
```diff
@@ -1012,6 +1004,26 @@ impl<'tcx> GotocCtx<'tcx> {
         Stmt::block(vec![src_align_check, dst_align_check, overflow_check, copy_expr], loc)
     }
 
+    // In some contexts (e.g., compilation-time evaluation),
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1007]: Kudos to @adpaco-aws. :)

File: kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs [hunk 2]
```diff
@@ -122,11 +122,9 @@ impl<'tcx> GotocCtx<'tcx> {
             }
             StatementKind::StorageLive(_) => Stmt::skip(location), // TODO: fix me
             StatementKind::StorageDead(_) => Stmt::skip(location), // TODO: fix me
-            StatementKind::CopyNonOverlapping(box mir::CopyNonOverlapping {
-                ref src,
-                ref dst,
-                ref count,
-            }) => {
+            mir::StatementKind::Intrinsic(box NonDivergingIntrinsic::CopyNonOverlapping(
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs:125]: Probably not. Let me remove that.

File: cprover_bindings/src/goto_program/expr.rs [hunk 3]
```diff
@@ -544,6 +544,13 @@ impl Expr {
         Self::double_constant(c)
     }
 
+    /// `if (c) { t } else { e }`
+    pub fn if_then_else_expr(c: Expr, t: Expr, e: Expr) -> Self {
```
[the remark this answers is not in the document]
Comment (celinval) [reply cprover_bindings/src/goto_program/expr.rs:548]: Good call! Let me replace that.

[the remark this answers is not in the document]
Comment (adpaco-aws) [reply cprover_bindings/src/goto_program/expr.rs:548]: Mea culpa, I wasn't able to find it.

**Review by zhassan-aws — APPROVED** [review]
Thanks!

**Pull request merged**